velox/connectors/Connector.h

```diff
@@ -287,6 +287,67 @@ class DataSource {
   }
 };
 
+/// Represents a lookup request for a given input.
+struct LookupRequest {
+  /// Contains the input column vectors used by lookup join and range
+  /// conditions.
+  RowVectorPtr input{nullptr};
+
+  LookupRequest(RowVectorPtr input) : input(std::move(input)) {}
+};
+
+/// Represents the lookup result for a subset of input produced by the
+/// 'LookupResultIterator'.
+struct LookupResult {
+  /// Used to specify the match result for a subset of input rows. It provides
+  /// an array of indices of the matched output rows in 'lookupOutput'. Each
+  /// array element contains all the matched rows for a given input row. If it
+  /// is empty, then there is no match for the given input row. 'inputHits'
+  /// contains the input indices in the same order of the corresponding input
+  /// rows in the lookup request. 'LookupResultIterator' also produces the
+  /// lookup result in the input request row order, and produces all the output
+  /// matches for a given input row in one lookup result.
+  ArrayVectorPtr inputHits;
+
+  /// The output vector that contains the lookup result rows.
+  RowVectorPtr lookupOutput{nullptr};
+
+  LookupResult(ArrayVectorPtr _inputHits, RowVectorPtr _lookupOutput)
+      : inputHits(std::move(_inputHits)),
+        lookupOutput(std::move(_lookupOutput)) {
+    VELOX_CHECK_EQ(
+        inputHits->elements()->typeKind(),
+        TypeKind::INTEGER,
```

Comment:
Why do we need an ArrayVector? Can we not use a simple std::vector or BufferPtr?

Given input of 10 rows, if all rows have 1 hit we get: 0, 1, 2, 3, 4, ....9

If every other row has 2 hits: 0, 0, 2, 2, 4, 4, 6, 6,...

Essentially, we provide a mapping from output row index to input row index.